testing/dart/image_filter_test.dart

```diff
@@ -0,0 +1,125 @@
+// Copyright 2019 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:typed_data';
+import 'dart:ui';
+
+import 'package:test/test.dart';
+
+const Color green = Color(0xFF00AA00);
+
+const int greenCenterBlurred = 0x1C001300;
+const int greenSideBlurred = 0x15000E00;
+const int greenCornerBlurred = 0x10000A00;
+
+const int greenCenterScaled = 0xFF00AA00;
+const int greenSideScaled = 0x80005500;
+const int greenCornerScaled = 0x40002B00;
+
+void main() {
+  Future<Uint32List> getBytesForPaint(Paint paint, {int width = 3, int height = 3}) async {
+    final PictureRecorder recorder = PictureRecorder();
+    final Canvas recorderCanvas = Canvas(recorder);
+    recorderCanvas.drawRect(const Rect.fromLTRB(1.0, 1.0, 2.0, 2.0), paint);
+    final Picture picture = recorder.endRecording();
+    final Image image = await picture.toImage(width, height);
+    final ByteData bytes = await image.toByteData();
+
+    expect(bytes.lengthInBytes, width * height * 4);
+    return bytes.buffer.asUint32List();
+  }
+
+  ImageFilter makeBlur(double sigmaX, double sigmaY) =>
+    ImageFilter.blur(sigmaX: sigmaX, sigmaY: sigmaY);
+
+  ImageFilter makeScale(double scX, double scY,
+                        [double trX = 0.0, double trY = 0.0,
+                         FilterQuality quality = FilterQuality.low]) {
+    trX *= 1.0 - scX;
+    trY *= 1.0 - scY;
+    return ImageFilter.matrix(Float64List.fromList(<double>[
+      scX, 0.0, 0.0, 0.0,
+      0.0, scY, 0.0, 0.0,
+      0.0, 0.0, 1.0, 0.0,
+      trX, trY, 0.0, 1.0,
+    ]), filterQuality: quality);
+  }
+
+  List<ImageFilter> makeList() {
+    return <ImageFilter>[
+      makeBlur(10.0, 10.0),
+      makeBlur(10.0, 20.0),
+      makeBlur(20.0, 20.0),
+      makeScale(10.0, 10.0),
+      makeScale(10.0, 20.0),
+      makeScale(20.0, 10.0),
+      makeScale(10.0, 10.0, 1.0, 1.0),
+      makeScale(10.0, 10.0, 0.0, 0.0, FilterQuality.medium),
+      makeScale(10.0, 10.0, 0.0, 0.0, FilterQuality.high),
+      makeScale(10.0, 10.0, 0.0, 0.0, FilterQuality.none),
+    ];
+  }
+
+  void checkEquality(List<ImageFilter> a, List<ImageFilter> b) {
+    for (int i = 0; i < a.length; i++) {
+      for(int j = 0; j < a.length; j++) {
+        if (i == j) {
+          expect(a[i], b[j]);
```

Comment:
I find `expect(a[i], equals(b[j]));` to be a little more clearer than `expect(a[i], b[j]);` although they're equivalent.

Reply:
I can see that.
I could also add a comment in the "else" section that says that I am deliberately not testing the hashcodes because their API contract allows them to be equal when the objects are not equal.

Reply:
Coming in the next commit.